File: src/scrub.zig
const std = @import("std");
const helpers = @import("helpers.zig");
const state_mod = @import("state.zig");
const replication = @import("replication.zig");
const c = helpers.c;
const log = helpers.log;
const FsState = state_mod.FsState;

pub fn scrubLoop(state: *FsState) void {
    if (shouldScrubImmediately(state)) {
        log.info("scrub overdue, running immediately", .{});
        runScrub(state);
    }

    while (!state.shutdown.load(.acquire)) {
        const sleep_ns = nsUntilNextScrub(state.scrub_hour, state.scrub_minute);
        var remaining = sleep_ns;
        while (remaining > 0 and !state.shutdown.load(.acquire)) {
            const chunk = @min(remaining, 1_000_000_000);
            std.Thread.sleep(chunk);
            remaining -= chunk;
        }
        if (state.shutdown.load(.acquire)) break;
        runScrub(state);
    }
}

pub fn shouldScrubImmediately(state: *FsState) bool {
    const ts_path = std.fs.path.join(state.allocator, &.{ state.backing_dir, ".helmetfs", "scrub.timestamp" }) catch return true;
    defer state.allocator.free(ts_path);

    const file = std.fs.openFileAbsolute(ts_path, .{}) catch return true;
    defer file.close();

    var buf: [64]u8 = undefined;
    const n = file.read(&buf) catch return true;
    const content = std.mem.trimRight(u8, buf[0..n], "\n\r ");
    const last_scrub = std.fmt.parseInt(i64, content, 10) catch return true;

    const now = std.time.timestamp();
    return (now - last_scrub) > 86400;
}

pub fn nsUntilNextScrub(target_hour: u8, target_minute: u8) u64 {
    const now_ts = std.time.timestamp();
    var time_val: c.time_t = @intCast(now_ts);
    var tm: c.struct_tm = undefined;
    _ = c.localtime_r(&time_val, &tm);

    const local_day_sec: u64 = @intCast(@as(i64, tm.tm_hour) * 3600 + @as(i64, tm.tm_min) * 60 + @as(i64, tm.tm_sec));
    const target_sec: u64 = @as(u64, target_hour) * 3600 + @as(u64, target_minute) * 60;

    const secs_until = if (target_sec > local_day_sec)
        target_sec - local_day_sec
    else
        86400 - local_day_sec + target_sec;

    return secs_until * 1_000_000_000;
}

pub fn runScrub(state: *FsState) void {
    log.info("starting scrub", .{});
    const start_ms = std.time.milliTimestamp();
    var files_checked: u64 = 0;
    var corruptions_found: u64 = 0;
    var repairs: u64 = 0;

    var dir = std.fs.openDirAbsolute(state.backing_dir, .{ .iterate = true }) catch |err| {
        log.err("scrub: failed to open backing dir: {}", .{err});
        return;
    };
    defer dir.close();

    var walker = dir.walk(state.allocator) catch |err| {
        log.err("scrub: failed to walk backing dir: {}", .{err});
        return;
    };
    defer walker.deinit();

    while (walker.next() catch null) |entry| {
        if (state.shutdown.load(.acquire)) break;
        if (entry.kind == .directory) continue;
        if (entry.kind == .sym_link) continue;
        if (std.mem.startsWith(u8, entry.path, ".helmetfs")) continue;
        if (std.mem.endsWith(u8, entry.path, ".sum")) continue;

        const rel_path = state.allocator.dupe(u8, entry.path) catch continue;
        defer state.allocator.free(rel_path);

        if (state.path_state.hasWriteRef(rel_path)) continue;

        scrubFile(state, rel_path, &corruptions_found, &repairs) catch |err| {
            log.err("scrub: error checking {s}: {}", .{ rel_path, err });
        };
        files_checked += 1;
    }

    const end = std.time.timestamp();
    const end_ms = std.time.milliTimestamp();
    const duration_ms: u64 = @intCast(end_ms - start_ms);

    writeScrubTimestamp(state, end);

    log.info("scrub complete: checked={d}, corruptions={d}, repairs={d}, duration={d}ms", .{
        files_checked, corruptions_found, repairs, duration_ms,
    });
}

pub fn scrubFile(state: *FsState, rel_path: []const u8, corruptions: *u64, repairs_count: *u64) !void {
    const backing_path = try std.fs.path.join(state.allocator, &.{ state.backing_dir, rel_path });
    defer state.allocator.free(backing_path);
    const sum_path = try std.fmt.allocPrint(state.allocator, "{s}.sum", .{backing_path});
    defer state.allocator.free(sum_path);

    const current_hex = helpers.computeBlake3(backing_path) catch |err| {
        log.err("scrub: failed to compute checksum for {s}: {}", .{ rel_path, err });
        return err;
    };

    const stored_hex = helpers.readSumFile(state.allocator, sum_path) catch |err| switch (err) {
        error.FileNotFound => {
            log.info("scrub: adopting untracked file {s}", .{rel_path});
            helpers.writeSumFile(sum_path, &current_hex) catch |we| {
                log.err("scrub: failed to write .sum for {s}: {}", .{ rel_path, we });
                return we;
            };
            state.repl_log.enqueue(.put, rel_path) catch |enq_err| {
                log.err("scrub: failed to enqueue replication for {s}: {}", .{ rel_path, enq_err });
            };
            return;
        },
        else => return err,
    };
    defer state.allocator.free(stored_hex);

    if (std.mem.eql(u8, &current_hex, stored_hex)) {
        return;
    }

    corruptions.* += 1;
    log.warn("scrub: CORRUPTION detected in {s}", .{rel_path});

    const has_pending = state.repl_log.hasPendingPut(rel_path);

    const replica_path = try std.fs.path.join(state.allocator, &.{ state.replica_dir, "files", rel_path });
    defer state.allocator.free(replica_path);
    const replica_sum_path = try std.fmt.allocPrint(state.allocator, "{s}.sum", .{replica_path});
    defer state.allocator.free(replica_sum_path);

    const replica_hex = helpers.readSumFile(state.allocator, replica_sum_path) catch {
        log.err("scrub: replica unavailable for repair of {s}", .{rel_path});
        return;
    };
    defer state.allocator.free(replica_hex);

    const replica_computed = helpers.computeBlake3(replica_path) catch {
        log.err("scrub: cannot read replica file for repair of {s}", .{rel_path});
        return;
    };

    if (!std.mem.eql(u8, &replica_computed, replica_hex)) {
        log.warn("scrub: replica also corrupt for {s}, cannot repair", .{rel_path});
        return;
    }

    if (has_pending) {
        log.warn("scrub: skipping repair of {s} — pending replication means replica is stale", .{rel_path});
        return;
    }

    // Re-check write ref before overwriting — a writer may have opened the
    // file between the initial check and now.
    if (state.path_state.hasWriteRef(rel_path)) {
        log.info("scrub: skipping repair of {s} — file now has open writer", .{rel_path});
        return;
    }

    // Skip if dirty — write completed but hasn't been checksummed yet.
    // (See TLA+ ScrubRepair precondition: dirty_gen = clean_gen.)
    if (state.path_state.isDirty(rel_path)) {
        log.info("scrub: skipping repair of {s} — file is dirty (pending checksum)", .{rel_path});
        return;
    }

    log.info("scrub: repairing {s} from replica", .{rel_path});

    replication.copyFileWithSync(replica_path, backing_path) catch |err| {
        log.err("scrub: failed to repair {s}: {}", .{ rel_path, err });
        return;
    };
    helpers.writeSumFile(sum_path, &replica_computed) catch |err| {
        log.err("scrub: failed to write .sum after repair of {s}: {}", .{ rel_path, err });
        return;
    };
    repairs_count.* += 1;
    log.info("scrub: successfully repaired {s}", .{rel_path});
}

pub fn writeScrubTimestamp(state: *FsState, ts: i64) void {
    const ts_path = std.fs.path.join(state.allocator, &.{ state.backing_dir, ".helmetfs", "scrub.timestamp" }) catch return;
    defer state.allocator.free(ts_path);
    const file = std.fs.createFileAbsolute(ts_path, .{}) catch return;
    defer file.close();
    const ts_str = std.fmt.allocPrint(state.allocator, "{d}\n", .{ts}) catch return;
    defer state.allocator.free(ts_str);
    file.writeAll(ts_str) catch return;
    file.sync() catch return;
}

test "nsUntilNextScrub returns value in (0, 24h]" {
    const testing = std.testing;
    for ([_]u8{ 0, 3, 6, 12, 18, 23 }) |hour| {
        const ns = nsUntilNextScrub(hour, 0);
        try testing.expect(ns > 0);
        try testing.expect(ns <= 86400 * 1_000_000_000);
    }
}

File: src/scrub.rs
//! Scrub: periodic integrity checking and self-healing.
//!
//! The scrub walks all files in the backing directory, and for each file:
//!
//! 1. **Untracked** (no `.sum` sidecar): compute checksum, write `.sum`,
//!    enqueue a `put` to replicate.
//!
//! 2. **Clean** (has `.sum`, checksum matches): nothing to do.
//!
//! 3. **Corrupt** (has `.sum`, checksum mismatch): attempt self-healing from
//!    the replica — but only if:
//!    - no pending `put` in the replication log for this path,
//!    - no open write references,
//!    - not dirty,
//!    - the replica's file checksum matches the replica's `.sum`.
//!
//! Scrub timing: on startup if no `scrub.timestamp` or if it's older than
//! `scrub_interval_secs`. After each scrub completes, update the timestamp.

use std::fs;
use std::io::Write;
use std::path::Path;
use std::sync::atomic::Ordering;
use std::sync::Arc;
use std::time::{Duration, SystemTime};

use crate::helpers;
use crate::replication::copy_file_with_sync;
use crate::state::{self, checksum_and_enqueue, FsState};

/// Run the scrub thread. Performs an initial scrub if needed, then sleeps for
/// `scrub_interval_secs` between runs.
pub fn scrub_thread(state: &Arc<FsState>) {
    log::info!("Scrub thread started");
    loop {
        if state.shutting_down.load(Ordering::Relaxed) {
            break;
        }

        if should_scrub(state) {
            log::info!("Starting scrub");
            let (checked, healed, adopted) = run_scrub(state);
            log::info!(
                "Scrub complete: checked={}, healed={}, adopted={}",
                checked,
                healed,
                adopted
            );
            update_scrub_timestamp(state);
        }

        // Sleep in small increments so we can respond to shutdown.
        // Use wall-clock target to avoid drift from scrub execution time.
        let target = std::time::Instant::now() + Duration::from_secs(state.scrub_interval_secs);
        while std::time::Instant::now() < target {
            if state.shutting_down.load(Ordering::Relaxed) {
                break;
            }
            std::thread::sleep(Duration::from_secs(1));
        }
    }
    log::info!("Scrub thread stopped");
}

/// Check if a scrub should run now.
fn should_scrub(state: &FsState) -> bool {
    let ts_path = state.backing_dir.join(".helmetfs").join("scrub.timestamp");
    match fs::metadata(&ts_path) {
        Ok(meta) => {
            let modified = meta
                .modified()
                .unwrap_or(SystemTime::UNIX_EPOCH);
            let elapsed = SystemTime::now()
                .duration_since(modified)
                .unwrap_or(Duration::from_secs(u64::MAX));
            elapsed.as_secs() >= state.scrub_interval_secs
        }
        Err(_) => true, // No timestamp file — scrub immediately
    }
}

/// Update the scrub timestamp file.
fn update_scrub_timestamp(state: &FsState) {
    let ts_path = state.backing_dir.join(".helmetfs").join("scrub.timestamp");
    if let Ok(mut f) = fs::File::create(&ts_path) {
        let _ = writeln!(f, "{}", humantime(SystemTime::now()));
    }
}

fn humantime(t: SystemTime) -> String {
    match t.duration_since(SystemTime::UNIX_EPOCH) {
        Ok(d) => d.as_secs().to_string(),
        Err(_) => "0".to_string(),
    }
}

/// Walk the backing directory and scrub all regular files.
/// Returns (checked, healed, adopted).
fn run_scrub(state: &FsState) -> (u64, u64, u64) {
    let mut checked = 0u64;
    let mut healed = 0u64;
    let mut adopted = 0u64;

    walk_dir(state, &state.backing_dir, &mut checked, &mut healed, &mut adopted);

    (checked, healed, adopted)
}

fn walk_dir(
    state: &FsState,
    dir: &Path,
    checked: &mut u64,
    healed: &mut u64,
    adopted: &mut u64,
) {
    let entries = match fs::read_dir(dir) {
        Ok(e) => e,
        Err(e) => {
            log::warn!("Scrub: failed to read dir {:?}: {}", dir, e);
            return;
        }
    };

    for entry in entries {
        if state.shutting_down.load(Ordering::Relaxed) {
            return;
        }

        let entry = match entry {
            Ok(e) => e,
            Err(_) => continue,
        };

        let path = entry.path();

        // Skip .helmetfs directory
        if let Ok(rel) = path.strip_prefix(&state.backing_dir) {
            if rel.starts_with(".helmetfs") {
                continue;
            }
        }

        let ft = match entry.file_type() {
            Ok(ft) => ft,
            Err(_) => continue,
        };

        if ft.is_dir() {
            walk_dir(state, &path, checked, healed, adopted);
            continue;
        }

        if !ft.is_file() {
            // Symlinks don't have checksums; skip
            continue;
        }

        // Skip .sum files themselves
        if path.extension().and_then(|e| e.to_str()) == Some("sum") {
            continue;
        }

        // .sum.tmp files — skip
        if path.to_str().map_or(false, |s| s.ends_with(".sum.tmp")) {
            continue;
        }

        let rel = match path.strip_prefix(&state.backing_dir) {
            Ok(r) => r.to_string_lossy().to_string(),
            Err(_) => continue,
        };

        *checked += 1;

        let sum_path = helpers::sum_path_for(&path);

        if !sum_path.exists() {
            // Untracked file — adopt it
            log::info!("Scrub: adopting untracked file: {}", rel);
            checksum_and_enqueue(state, &rel);
            *adopted += 1;
            continue;
        }

        // Has .sum — verify checksum
        let stored_sum = match helpers::read_sum_file(&sum_path) {
            Ok(s) => s,
            Err(e) => {
                log::warn!("Scrub: failed to read .sum for {}: {}", rel, e);
                continue;
            }
        };

        let actual_sum = match helpers::compute_blake3(&path) {
            Ok(s) => s,
            Err(e) => {
                log::warn!("Scrub: failed to checksum {}: {}", rel, e);
                continue;
            }
        };

        if actual_sum == stored_sum {
            // Clean — nothing to do
            continue;
        }

        // Corrupt — attempt self-healing
        log::warn!("Scrub: corruption detected in {}", rel);

        if !can_heal_file(state, &rel) {
            log::warn!("Scrub: cannot heal {} (busy or pending replication)", rel);
            continue;
        }

        if try_heal(state, &rel, &path) {
            log::info!("Scrub: healed {}", rel);
            *healed += 1;
        } else {
            log::error!("Scrub: failed to heal {}", rel);
        }
    }
}

/// Check if a file is eligible for healing.
fn can_heal_file(state: &FsState, rel: &str) -> bool {
    state::can_heal(state.repl_log.has_pending_put(rel), state.is_busy(rel))
}

/// Attempt to heal a corrupted file from the replica.
/// Returns true on success.
fn try_heal(state: &FsState, rel: &str, backing_path: &Path) -> bool {
    let replica_path = state.replica_file_path(rel);
    let replica_sum_path = helpers::sum_path_for(&replica_path);

    // Replica file must exist
    if !replica_path.exists() {
        log::warn!("Scrub: no replica copy for {}", rel);
        return false;
    }

    // Replica .sum must exist
    let replica_sum = match helpers::read_sum_file(&replica_sum_path) {
        Ok(s) => s,
        Err(_) => {
            log::warn!("Scrub: no replica .sum for {}", rel);
            return false;
        }
    };

    // Verify replica integrity: replica file checksum must match replica .sum
    let replica_actual = match helpers::compute_blake3(&replica_path) {
        Ok(s) => s,
        Err(e) => {
            log::warn!("Scrub: failed to checksum replica for {}: {}", rel, e);
            return false;
        }
    };

    if replica_actual != replica_sum {
        log::warn!("Scrub: replica itself is corrupt for {}", rel);
        return false;
    }

    // Restore from replica
    if let Err(e) = copy_file_with_sync(&replica_path, backing_path) {
        log::error!("Scrub: failed to restore {} from replica: {}", rel, e);
        return false;
    }

    // Also restore the .sum
    let backing_sum_path = helpers::sum_path_for(backing_path);
    if let Err(e) = copy_file_with_sync(&replica_sum_path, &backing_sum_path) {
        log::error!("Scrub: failed to restore .sum for {}: {}", rel, e);
        // File was restored even if .sum copy failed, still count as healed
    }

    true
}
